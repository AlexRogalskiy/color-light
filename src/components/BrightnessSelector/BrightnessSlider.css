.brightness-slider {
  width: auto;
}

.brightness-slider .ant-slider-rail {
  background: linear-gradient(to right, #444, white);
}

.brightness-slider .ant-slider-track {
  opacity: 0;
}
