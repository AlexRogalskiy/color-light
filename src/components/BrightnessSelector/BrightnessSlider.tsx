import React from 'react'

import './BrightnessSlider.css'

import { Slider } from 'antd'

interface IBrightnessSliderProps {
  brightness: number
  setBrightness: (brightness: number) => void
}

const BrightnessSlider = ({ brightness, setBrightness }: IBrightnessSliderProps) => {
  // When the user is moving the slider, we overwrite the brightness using this state variable until
  // the user releases the handle. It will be null when the value is not overwritten.
  const [overwrittenBrightness, setOverwrittenBrightness] = React.useState<number | null>(null)

  return (
    <div data-test-id={'brightness-slider'}>
      <Slider
        step={0.01}
        tipFormatter={(value) => `${Math.round((value ?? 0) * 100)}%`}
        max={1}
        value={overwrittenBrightness ?? brightness}
        onChange={(value: any) => typeof value === 'number' && setOverwrittenBrightness(value)}
        onAfterChange={(value: any) => {
          // Stop overwriting the brightness
          setOverwrittenBrightness(null)

          // Update the brightness
          typeof value === 'number' && setBrightness(value)
        }}
        className={'brightness-slider'}
      />
    </div>
  )
}

export default BrightnessSlider
