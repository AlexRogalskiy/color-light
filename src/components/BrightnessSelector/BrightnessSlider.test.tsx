import React from 'react'
import { cleanup, fireEvent, render, queryByRole } from '@testing-library/react'
import BrightnessSlider from './BrightnessSlider'

// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup)

describe('<BrigtnessSlider />', () => {
  it('Renders the slider', () => {
    const { container } = render(<BrightnessSlider brightness={100} setBrightness={jest.fn()} />)

    expect(container).toMatchSnapshot()
  })

  it('Emits setBrightness update on mouse select', () => {
    // Mock setBrightness
    const setBrightness = jest.fn()

    const { getByRole } = render(<BrightnessSlider brightness={1} setBrightness={setBrightness} />)

    const slider = getByRole(/slider/i)

    fireEvent.mouseDown(slider)
    fireEvent.mouseUp(slider)

    expect(setBrightness).toBeCalledWith(0)
  })
})
