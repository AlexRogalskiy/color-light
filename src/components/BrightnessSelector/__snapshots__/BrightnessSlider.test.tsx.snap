// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BrigtnessSlider /> Renders the slider 1`] = `
<div>
  <div
    data-test-id="brightness-slider"
  >
    <div
      class="ant-slider brightness-slider"
    >
      <div
        class="ant-slider-rail"
      />
      <div
        class="ant-slider-track"
        style="left: 0%; width: 100%;"
      />
      <div
        class="ant-slider-step"
      />
      <div
        aria-disabled="false"
        aria-valuemax="1"
        aria-valuemin="0"
        aria-valuenow="1"
        class="ant-slider-handle"
        role="slider"
        style="left: 100%; transform: translateX(-50%);"
        tabindex="0"
      />
      <div
        class="ant-slider-mark"
      />
    </div>
  </div>
</div>
`;
